### PR TITLE
Improve the register_index_accessor Documentation

### LINKED ISSUE #49202: DOC: Improve the register_index_accessor Documentation
### Pandas version checks

- [X] I have checked that the issue still exists on the latest versions of the docs on `main` [here](https://pandas.pydata.org/docs/dev/)

### Location of the documentation

https://pandas.pydata.org/docs/reference/api/pandas.api.extensions.register_index_accessor.html#pandas.api.extensions.register_index_accessor

### Documentation problem

There is no example showing how to create or what is required on the class to make an index accessor.

### Suggested fix for documentation

Add a link to an example or expand the topic to include a how to implement this.  Also document what is required to make a custom indexer.

### PR DESCRIPTION
**Pull request by kanpuriyanawab**
- [x] closes #49202 
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by shivance** [comment]
Some doc checks are failing, here is the issue... 
_register_accessor has it's own docs, which are appearing for all of 3 index, series and dataframe accessor.
Now that I've put in the correct docs of this function.. the old docs are still being injected apart from the new docs.

How to remove this issue ? .. My guess is to remove the decorator itself for this `_register_index_accessor` method.


I'll take a look by doing this change and see if doc test pass.

File: pandas/core/accessor.py [hunk 1]
```diff
@@ -293,6 +293,66 @@ def register_series_accessor(name):
 
 @doc(_register_accessor, klass="Index")
```
Comment (phofl) [inline pandas/core/accessor.py:294]: I think removing this should fix the doc build

pre-commit is failing, you can run it locally before comitting

**Comment by shivance** [comment]
@phofl The CI is all clean now, and I think PR is ready for another round of review .

**Review by phofl — COMMENTED** [review]
Can you post a screenshot on how this looks?

**Comment by phofl** [comment]
Can you rebuild our extensions?

**Comment by shivance** [comment]
> Can you rebuild our extensions?

how to rebuilding extensions ? 

I tried reinstalling anaconda, as well as just python in windows machine, facing the same issue @phofl 
python version 3.8.10

**Comment by phofl** [comment]
https://pandas.pydata.org/docs/development/contributing_environment.html

The first note about building the c extensions

**Comment by shivance** [comment]
image

@phofl  Docker is amazing, managed to pull it and build documentation. Here is how it looks like

**Comment by shivance** [comment]
Is this PR ready to merge @phofl ?

**Comment by shivance** [comment]
@mroeschke @jbrockmendel could you please review this PR and give feedback?

File: pandas/core/accessor.py [hunk 2]
```diff
@@ -291,8 +291,65 @@ def register_series_accessor(name):
     return _register_accessor(name, Series)
 
 
-@doc(_register_accessor, klass="Index")
```
Comment (mroeschke) [inline pandas/core/accessor.py:294]: So any documentation changes should be made to the docstring of `_register_accessor` as it is shared with 3 other methods

**Review by mroeschke — CHANGES_REQUESTED** [review]

File: pandas/core/accessor.py (hunk 1 above)
Comment (shivance) [reply pandas/core/accessor.py]: @mroeschke please see first comment

File: pandas/core/accessor.py (hunk 2 above)
Comment (shivance) [reply pandas/core/accessor.py:294]: As I removed the decorator, the docstring is no longer shared

I believe it was shared earlier just to make the class visible in docs, but shared documentation wasn't right

Comment (mroeschke) [reply pandas/core/accessor.py:294]: I think based on the issue suggestion:

> Add a link to an example or expand the topic to include a how to implement this. Also document what is required to make a custom indexer.

all 3 functions that share this docstring should be updated

Comment (shivance) [reply pandas/core/accessor.py:294]: So shall I make appropriate docstring for other accessors as well?

Again the decorator will have to be removed as they won't share the docstrings anymore.

Comment (mroeschke) [reply pandas/core/accessor.py:294]: I don't think the decorator needs to be removed. For example

> Also document what is required to make a custom indexer.

Should be the same and described for Series, DataFrame, and Index so the docstring should be shared

Comment (shivance) [reply pandas/core/accessor.py:294]: > @mroeschke Some doc checks are failing, here is the issue... _register_accessor has it's own docs, which are appearing for all of 3 index, series and dataframe accessor. Now that I've put in the correct docs of this function.. the old docs are still being injected apart from the new docs.
> 
> How to remove this issue ? .. My guess is to remove the decorator itself for this `_register_index_accessor` method.
> 
> What should be the correct way to fix the error?

@mroeschke I've already tried that. CI fails .

Comment (shivance) [reply pandas/core/accessor.py:294]: It detects dual docstring inserted, one due to decorator, and one, the modified string specific for register index accessor.

Comment (shivance) [reply pandas/core/accessor.py:294]: Please note that, series, dataframe and index accessor work differently, so their docs can't be same, can they?

File: pandas/core/accessor.py [hunk 3]
```diff
@@ -291,8 +291,65 @@ def register_series_accessor(name):
...
+        Name used when calling the accessor after its registered.
+
+    Returns
+    -------
+    callable
+        A class decorator.
+
+    See Also
+    --------
+    register_dataframe_accessor: Register a custom accessor on DataFrame objects.
```
Comment (jbrockmendel) [inline pandas/core/accessor.py:310]: i think supposed to have a space before the colon

File: pandas/core/accessor.py [hunk 4]
```diff
@@ -291,8 +291,65 @@ def register_series_accessor(name):
...
+    Notes
+    -----
+    When accessed, your accessor will be initialized with the pandas_obj object
+    the user is interacting with. The code signature must be::
+
+        def __init__(self, pandas_obj):
+            # constructor logic
+        ...
+
+    >>> pd.Series(['a', 'b']).dt
```
Comment (jbrockmendel) [inline pandas/core/accessor.py:322]: if you're making an Index-specific docstring, should you use an Index example here?

**Comment by mroeschke** [comment]
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.

**Pull request closed without merging**